pkg/objstore/caching/caching_bucket.go

```diff
@@ -0,0 +1,290 @@
+// cached blocks since last "fetch segment" is bigger than some N (minCachingBlocks). In this case,
+// previous "fetch segment" is finished, and new fetch segment will be started later, if needed.
+func (cb CachingBucket) createBlockProvider(ctx context.Context, name string, startOffset, endOffset int64, offsetKeyMap map[int64]string, blockHits map[string][]byte, lastBlockOffset int64, lastBlockLength int) blocksProvider {
+	perOffsetProvider := make(map[int64]blocksProvider, (endOffset-startOffset)/cb.chunkBlockSize)
+	allProviders := []blocksProvider(nil)
+
+	cachedBlocksProvider := newCachedBlocksProvider(offsetKeyMap, blockHits, cb.cachedBytes)
+	allProviders = append(allProviders, cachedBlocksProvider)
+
+	// Number of blocks in pending fetching and caching segments, starting from respective offsets.
+	fetchingBlocks, cachingBlocks := 0, 0
+	fetchingOffset, cachingOffset := int64(0), int64(0)
+
+	// Function that "flushes" pending segments, both fetching and caching.
+	// Fetching segment is created between fetchingOffset and fetchingOffsetEnd.
+	var flush = func(fetchingOffsetEnd int64) {
+		// we are going to use cached blocks, instead of fetching them.
+		bp := newFetchingBlocksProvider(ctx, cb.bucket, cb.cache, cb.chunkBlockSize, name, fetchingOffset, fetchingOffsetEnd, offsetKeyMap, lastBlockOffset, lastBlockLength, cb.chunkBlockTTL, cb.fetchedBytes)
+		allProviders = append(allProviders, bp)
+
+		for i := 0; i < fetchingBlocks; i++ {
+			perOffsetProvider[fetchingOffset+int64(i)*cb.chunkBlockSize] = bp
+		}
+
+		fetchingOffset = 0
+		fetchingBlocks = 0
+
+		for i := 0; i < cachingBlocks; i++ {
+			perOffsetProvider[cachingOffset+int64(i)*cb.chunkBlockSize] = cachedBlocksProvider
+		}
+		cachingBlocks = 0
+		cachingOffset = 0
+	}
+
+	for off := startOffset; off < endOffset; off += cb.chunkBlockSize {
+		if blockHits[offsetKeyMap[off]] != nil {
+			if fetchingBlocks == 0 {
+				// CachingBlocks must be 0 here.
+				perOffsetProvider[off] = cachedBlocksProvider
+			} else {
+				if cachingBlocks == 0 {
+					cachingOffset = off
+				}
+				cachingBlocks++
+				if int64(cachingBlocks)*cb.chunkBlockSize >= cb.minCachingBlocksSize {
+					flush(cachingOffset) // fetch everything up to cachingOffset
+				}
+			}
+		} else {
+			if fetchingBlocks == 0 {
+				fetchingOffset = off
+			}
+			// If there were any "caching" blocks, merge them into fetching segment.
+			fetchingBlocks += cachingBlocks
+			cachingBlocks = 0
+			fetchingBlocks++
+		}
+	}
+
+	if cachingBlocks > 0 {
+		flush(cachingOffset)
+	} else {
+		flush(endOffset)
+	}
+
+	// combination of caching and fetching block providers
+	return perOffsetBlockProvider{
+		blocks:  perOffsetProvider,
+		toClose: allProviders,
+	}
+}
+
+func cachingKeyObjectSize(name string) string {
+	return fmt.Sprintf("size:%s", name)
+}
+
+func cachingKeyObjectBlock(name string, start int64, end int64) string {
+	return fmt.Sprintf("block:%s:%d:%d", name, start, end)
+}
```

Comment:
I don't think it will be a problem here, where name looks like `/01E1TR0Q8MKHM3BC3JX1JE6DP3/chunks/000007` (40 bytes) plus two offsets in 512 MiB file (`536870756:536870756` is 19 bytes), but it's something to keep in mind.